`typeinfo` fails with member variables of user-defined type

The following Spicy code gets a C++ compiler error:

```
module Test;

type S = struct {};

type Data = unit {
    var y: S;
    var tyinfo: type = typeinfo(self.y);
};
```

Error:

```
$ spicyc -j test.spicy
/private/var/folders/w6/h7tg594x0yng0_yzvcsn6mv00000gn/T/Test_8e15d3a6857f8e86-946611703c79621d.cc:59:19: error: use of undeclared identifier 'type_info'; did you mean 'std::type_info'?
   59 |         tyinfo = &type_info::__ti_Test__S_namex2aTest__Sx2b;
      |                   ^~~~~~~~~
      |                   std::type_info
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.2.sdk/usr/include/c++/v1/cxxabi.h:32:7: note: 'std::type_info' declared here
   32 | class type_info; // forward declaration
      |       ^
/private/var/folders/w6/h7tg594x0yng0_yzvcsn6mv00000gn/T/Test_8e15d3a6857f8e86-946611703c79621d.cc:59:30: error: no member named '__ti_Test__S_namex2aTest__Sx2b' in 'std::type_info'
   59 |         tyinfo = &type_info::__ti_Test__S_namex2aTest__Sx2b;
      |                   ~~~~~~~~~~~^
2 errors generated.
[error] spicyc: JIT compilation failed
```

I tested and this errors if `y` is `struct`, `unit`, or `enum` type. There are possibly more

--- a/hilti/toolchain/include/compiler/detail/cxx/unit.h
+++ b/hilti/toolchain/include/compiler/detail/cxx/unit.h
@@ -105,16 +105,17 @@ protected:
 
 private:
     enum class Phase {
-        Includes,
-        Forwards,
-        Enums,
-        Types,
-        PublicAliases,
         Constants,
-        Globals,
+        Enums,
+        Forwards,
         Functions,
+        Globals,
+        Implementations,
+        Includes,
+        PublicAliases,
+        TypeInfoForwards,
         TypeInfos,
-        Implementations
+        Types,
     };
 
     using cxxDeclaration = std::variant<declaration::IncludeFile, declaration::Global, declaration::Constant,

--- a/hilti/toolchain/src/compiler/cxx/unit.cc
+++ b/hilti/toolchain/src/compiler/cxx/unit.cc
@@ -124,9 +124,13 @@ void Unit::_emitDeclarations(const cxxDeclaration& decl, Formatter& f, Phase pha
 
         void operator()(const declaration::Constant& d) {
             if ( isTypeInfo(d.id) ) {
-                if ( phase == Phase::TypeInfos ) {
-                    // We split these out because creating the type information
-                    // needs access to all other types.
+                // We split these out because creating the type information
+                // needs access to all other types.
+                if ( phase == Phase::TypeInfoForwards && d.linkage == "extern" ) {
+                    f << d;
+                    return;
+                }
+                else if ( phase == Phase::TypeInfos && d.linkage != "extern" ) {
                     f << d;
                     return;
                 }
@@ -198,8 +202,9 @@ void Unit::_emitDeclarations(const cxxDeclaration& decl, Formatter& f, Phase pha
 }
 
 void Unit::_generateCode(Formatter& f, bool prototypes_only, bool include_all_implementations) {
-    const Phase phases[] = {Phase::Forwards,      Phase::Enums,   Phase::Types,     Phase::Constants,
-                            Phase::PublicAliases, Phase::Globals, Phase::Functions, Phase::TypeInfos};
+    const Phase phases[] = {Phase::TypeInfoForwards, Phase::Forwards,      Phase::Enums,   Phase::Types,
+                            Phase::Constants,        Phase::PublicAliases, Phase::Globals, Phase::Functions,
+                            Phase::TypeInfos};
 
     for ( const auto& [_, decl] : _declarations )
         _emitDeclarations(decl, f, Phase::Includes, prototypes_only, include_all_implementations);

--- a/tests/Baseline/hilti.types.tuple.coercion/output
+++ b/tests/Baseline/hilti.types.tuple.coercion/output
@@ -8,6 +8,12 @@
 
 extern const char* __hlt_hlto_scope;
 
+namespace __hlt::type_info { namespace {
+    extern const ::hilti::rt::TypeInfo __ti_tuple_a__uint_64___b__string__tuplex2auint64_string_x2b;
+    extern const ::hilti::rt::TypeInfo __ti_optional_uint_64___optionalx2auint64x2b;
+    extern const ::hilti::rt::TypeInfo __ti_optional_string__optionalx2astringx2b;
+} }
+
 namespace __hlt::Foo {
     static auto f() -> ::hilti::rt::Tuple<::hilti::rt::integer::safe<uint64_t>, std::string>;
     extern void __init_module();
@@ -15,10 +21,7 @@ namespace __hlt::Foo {
 }
 
 namespace __hlt::type_info { namespace {
-    extern const ::hilti::rt::TypeInfo __ti_tuple_a__uint_64___b__string__tuplex2auint64_string_x2b;
-    extern const ::hilti::rt::TypeInfo __ti_optional_uint_64___optionalx2auint64x2b;
     const ::hilti::rt::TypeInfo __ti_optional_uint_64___optionalx2auint64x2b = { {}, "optional<uint<64>>", [](const void *self) { return hilti::rt::to_string(*reinterpret_cast<const std::optional<::hilti::rt::integer::safe<uint64_t>>*>(self)); }, new ::hilti::rt::type_info::Optional(&::hilti::rt::type_info::uint64, ::hilti::rt::type_info::Optional::accessor<::hilti::rt::integer::safe<uint64_t>>()) };
-    extern const ::hilti::rt::TypeInfo __ti_optional_string__optionalx2astringx2b;
     const ::hilti::rt::TypeInfo __ti_optional_string__optionalx2astringx2b = { {}, "optional<string>", [](const void *self) { return hilti::rt::to_string(*reinterpret_cast<const std::optional<std::string>*>(self)); }, new ::hilti::rt::type_info::Optional(&::hilti::rt::type_info::string, ::hilti::rt::type_info::Optional::accessor<std::string>()) };
     const ::hilti::rt::TypeInfo __ti_tuple_a__uint_64___b__string__tuplex2auint64_string_x2b = { {}, "tuple<a: uint<64>, b: string>", [](const void *self) { return hilti::rt::to_string(*reinterpret_cast<const ::hilti::rt::Tuple<::hilti::rt::integer::safe<uint64_t>, std::string>*>(self)); }, new ::hilti::rt::type_info::Tuple(std::vector<::hilti::rt::type_info::tuple::Element>({::hilti::rt::type_info::tuple::Element{ "a", &type_info::__ti_optional_uint_64___optionalx2auint64x2b, ::hilti::rt::tuple::elementOffset<::hilti::rt::Tuple<::hilti::rt::integer::safe<uint64_t>, std::string>, 0>() }, ::hilti::rt::type_info::tuple::Element{ "b", &type_info::__ti_optional_string__optionalx2astringx2b, ::hilti::rt::tuple::elementOffset<::hilti::rt::Tuple<::hilti::rt::integer::safe<uint64_t>, std::string>, 1>() }})) };
 } }

--- a/tests/spicy/types/type/type-info.spicy
+++ b/tests/spicy/types/type/type-info.spicy
@@ -10,6 +10,11 @@ type S = struct {
     b: int32;
 };
 
+type T = unit {
+    var x: S;
+    var y: type = typeinfo(self.x);
+};
+
 function f(ty: type) {
     print "in f:", ty;
 }

